README.rst
===========
NRPE README
===========

:Project page: https://github.com/KristianLyng/nrpe
:Author: Kristian Lyngstøl <kristian@bohemians.org>
:Date: March, 2011

.. contents::

.. sectnum::


New documentation
`````````````````

The following is added documentation for NRPE 3.0. It will change as the
project stabilize and approaches some sort of upstream-like state.

NRPE "3.0"
==========

This is a so far unofficial version of NRPE, dubbed 3.0 by yours truly. It
implements IPv6 support and does a general cleanup. The version-number 3.0
was chosen because it differs vastly from previous versions, but is subject
to change.

It is quite likely that several key features are broken at the moment, thus
any feedback is welcome. This applies particularly to non-standard configs.


About the IPv6 support
======================

NRPE "3.0" uses getaddrinfo() both for listening and connecting, which
makes it agnostic to IPv4 vs IPv6. However, it does present a challenge
with dual-stack: I have yet to implement support for listening to both IPv4
and IPv6. Connecting is no problem though: getaddrinfo() returns a list of
possible addresses to use and the correct way of using it is to iterate
over it until it connects. This will leave the choice of prioritization to
the operating system, which should return a sorted list.

An other flaw is that I have not verified how connection-filtering works at
the moment. This likely breaks for IPv6, thus you'll need to allow all,
then use firewall techniques. Frankly, this might be the Correct solution
anyway.

Build system
============

I've moved away from the "what the heck is this?"-system and to plain
autotools for "3.0". This should allow easier maintainability and makes
everything less magic. However, this is likely to have broken several key
features. I've only tested it on Debian (though the result has been tested
on other systems), but it shouldn't be that hard to adapt.

The general build process from scratch is::

        ./autogen.sh
        ./configure
        make
        make install # may require root. e.g: sudo make install

If you require a package, you can use::

        ./autogen.sh
        ./configure
        make dist

Which will build a proper tar-ball (which doesn't require you to run
``./autogen.sh``)

Dependencies
------------

To build, you will need:

- Basic build utilities (make, a C compiler/linker, etc)
- automake, autoconf, autoheader (i.e., autotools. This is only required if
  you do not have a tar-ball generated by ``make dist``)
- pkg-config and openssl if using ssl, or pass --without-ssl to
  ``./configure``

There are no dependencies to run the nrpe-server or use check_nrpe except
OpenSSL if you are using SSL.




Pre NRPE "3.0"-documentation
````````````````````````````


The following is the (reformatted) documentation that shipped with NRPE
2.*. It may not apply verbatim to the current version, but is left for
completeness.

It was written by Ethan Galstad (nagios@nagios.org).


Further documentation
=====================

For installation instructions and information on the design overview
of the NRPE addon, please read the PDF documentation that is found in
this directory: NRPE.pdf

Purpose
=======
The purpose of this addon is to allow you to execute Nagios
plugins on a remote host in as transparent a manner as possible.


Contents
========

There are two pieces to this addon:

1. NRPE
       - This program runs as a background process on the
         remote host and processes command execution requests
         from the check_nrpe plugin on the Nagios host.
         Upon receiving a plugin request from an authorized
         host, it will execute the command line associated
         with the command name it received and send the
         program output and return code back to the
         check_nrpe plugin

2. check_nrpe
         - This is a plugin that is run on the Nagios host
           and is used to contact the NRPE process on remote
           hosts.  The plugin requests that a plugin be
           executed on the remote host and wait for the NRPE
           process to execute the plugin and return the result.
           The plugin then uses the output and return code
           from the plugin execution on the remote host for
           its own output and return code.


Installing
==========

The check_nrpe plugin should be placed on the Nagios host along
with your other plugins.  In most cases, this will be in the
/usr/local/nagios/libexec directory.

The nrpe program and the configuration file (nrpe.cfg) should
be placed somewhere on the remote host.  Note that you will also
have to install some plugins on the remote host if you want to
make much use of this addon.



Configuring
===========

Sample config files for the NRPE daemon are located in the
sample-config/ subdirectory.


Running Under INETD or XINETD
=============================

If you plan on running nrpe under inetd or xinetd and making use
of TCP wrappers, you need to do the following things:


Update /etc/services
--------------------

Add a line to your /etc/services file as follows (modify the port
number as you see fit)::

	nrpe            5666/tcp	# NRPE

Add NRPE to inetd or xinetd configuration
-----------------------------------------

Add entries for the NRPE daemon to either your inetd or xinetd
configuration files.  Which one your use will depend on which
superserver is installed on your system.  Both methods are described
below.  NOTE: If you run nrpe under inetd or xinetd, the server_port
and allowed_hosts variables in the nrpe configuration file are
ignored.


INETD
.....

If your system uses the inetd superserver WITH tcpwrappers, add an
entry to /etc/inetd.conf as follows::

        nrpe 	stream 	tcp 	nowait 	<user> /usr/sbin/tcpd <nrpebin> -c <nrpecfg> --inetd

If your system uses the inetd superserver WITHOUT tcpwrappers, add an
entry to /etc/inetd.conf as follows::

	nrpe 	stream 	tcp 	nowait 	<user> <nrpebin> -c <nrpecfg> --inetd


- Replace <user> with the name of the user that the nrpe server should run as.
  Example: `nagios`
- Replace <nrpebin> with the path to the nrpe binary on your system.
  Example: `/usr/local/nagios/nrpe`
- Replace <nrpecfg> with the path to the nrpe config file on your system.
  Example: `/usr/local/nagios/nrpe.cfg`


XINETD
......

If your system uses xinetd instead of inetd, you'll probably
want to create a file called 'nrpe' in your /etc/xinetd.d
directory that contains the following entries::


	# default: on
	# description: NRPE
	service nrpe
	{
        	flags           = REUSE
	        socket_type     = stream
        	wait            = no
	        user            = <user>
        	server          = <nrpebin>
	        server_args     = -c <nrpecfg> --inetd
        	log_on_failure  += USERID
	        disable         = no
		only_from       = <ipaddress1> <ipaddress2> ...
	}


- Replace <user> with the name of the user that the nrpe server should run as.
- Replace <nrpebin> with the path to the nrpe binary on your system.
- Replace <nrpecfg> with the path to the nrpe config file on your system.
- Replace the <ipaddress> fields with the IP addresses of hosts which
  are allowed to connect to the NRPE daemon.  This only works if xinetd was
  compiled with support for tcpwrappers.

Restart inetd or xinetd
-----------------------

Restart inetd or xinetd will the following command (pick the
on that is appropriate for your system::

	/etc/rc.d/init.d/inet restart

	/etc/rc.d/init.d/xinetd restart

OpenBSD users can use the following command to restart inetd::

	kill -HUP `cat /var/run/inet.pid`

Update tcp-wrapper entries
--------------------------

Add entries to your /etc/hosts.allow and /etc/hosts.deny
file to enable TCP wrapper protection for the nrpe service.
This is optional, although highly recommended.

Configuring Things On The Nagios Host
=======================================

Examples for configuring the nrpe daemon are found in the sample
nrpe.cfg file included in this distribution.  That config file
resides on the remote host(s) along with the nrpe daemon.  The
check_nrpe plugin gets installed on the Nagios host.  In order
to use the check_nrpe plugin from within Nagios, you'll have
to define a few things in the host config file.  An example
command definition for the check_nrpe plugin would look like this::

        define command {
	        command_name	check_nrpe
	        command_line	/usr/local/nagios/libexec/check_nrpe -H $HOSTADDRESS$ -c $ARG1$
	}

In any service definitions that use the nrpe plugin/daemon to
get their results, you would set the service check command portion
of the definition to something like this (sample service definition
is simplified for this example)::

        define service{
                host_name		someremotehost
                service_description	someremoteservice
                check_command		check_nrpe!yourcommand
                ... etc ...
        }

where "yourcommand" is a name of a command that you define in
your nrpe.cfg file on the remote host (see the docs in the
sample nrpe.cfg file for more information).


Questions?
----------

If you have questions about this addon, or problems getting things
working, first try searching the nagios-users mailing list archives.
Details on searching the list archives can be found at
http://www.nagios.org

If all else fails, you can email me and I'll try and respond as
soon as I get a chance.

	-- Ethan Galstad (nagios@nagios.org)


debian/NEWS
nagios-nrpe (2.12-4) unstable; urgency=low

  The pidfile creation mechanism changed with this update. If you do not 
  add "pid_file=/var/run/nagios/nrpe.pid" to you nrpe config take care that
  the user "nagios" is able to write to your pidfile location. You can also 
  change the initscript to create the pid directory on your own. 

 -- Alexander Wirt <formorer@debian.org>  Tue, 07 Jul 2009 07:42:13 +0200

nagios-nrpe (2.12-3) unstable; urgency=low

  The homedirectory of the nagios user moved to /var/lib/nagios
  which is now common on all nagios related packages. Its recommended
  that you migrate an already existing nagios user to use /var/lib/nagios 
  as homedirectory. 

 -- Alexander Wirt <formorer@debian.org>  Sat, 21 Mar 2009 09:08:58 +0100

nagios-nrpe (2.4-1) unstable; urgency=low

  the nagios-nrpe-doc package is no longer provided.  the documentation
  can now be found in /usr/share/doc/nagios-nrpe-{server|plugins}.  new
  versions of the plugin and server packages conflict with the doc
  package to prevent the old (and possibly incorrect in the future)
  documentation from remaining.  to fully purge all information about
  the package you should run:
  	dpkg -P nagios-nrpe-doc

 -- sean finney <seanius@debian.org>  Mon, 13 Mar 2006 15:47:47 +0100


src/utils.c
/*
 * utils.c - NRPE Utility Functions
 *
 * Copyright (c) 1999-2006 Ethan Galstad (nagios@nagios.org)
 * Copyright (c) 2011 Kristian Lyngstol <kristian@bohemians.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

 /*
  * This file contains common network functions used in nrpe and check_nrpe.
  *
  * XXX: Much legacy or redundant. Fixing it as I go. -K
  *
  */

#include <arpa/inet.h>
#include <errno.h>
#include <netdb.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <time.h>
#include <unistd.h>
#include <stdarg.h>

#include "common.h"
#include "utils.h"

static unsigned long crc32_table[256];

/* build the crc table - must be called before calculating the crc value */
void generate_crc32_table(void)
{
	unsigned long crc, poly;
	int i, j;

	poly = 0xEDB88320L;
	for (i = 0; i < 256; i++) {
		crc = i;
		for (j = 8; j > 0; j--) {
			if (crc & 1)
				crc = (crc >> 1) ^ poly;
			else
				crc >>= 1;
		}
		crc32_table[i] = crc;
	}

	return;
}

/* calculates the CRC 32 value for a buffer */
unsigned long calculate_crc32(char *buffer, int buffer_size)
{
	register unsigned long crc;
	int this_char;
	int current_index;

	crc = 0xFFFFFFFF;

	for (current_index = 0; current_index < buffer_size; current_index++) {
		this_char = (int)buffer[current_index];
		crc =
		    ((crc >> 8) & 0x00FFFFFF) ^ crc32_table[(crc ^ this_char) &
							    0xFF];
	}

	return (crc ^ 0xFFFFFFFF);
}

/* fill a buffer with semi-random data */
void randomize_buffer(char *buffer, int buffer_size)
{
	FILE *fp;
	int x;
	int seed;

	/**** FILL BUFFER WITH RANDOM ALPHA-NUMERIC CHARACTERS ****/

	/***************************************************************
	   Only use alpha-numeric characters becase plugins usually
	   only generate numbers and letters in their output.  We
	   want the buffer to contain the same set of characters as
	   plugins, so its harder to distinguish where the real output
	   ends and the rest of the buffer (padded randomly) starts.
	***************************************************************/

	/* try to get seed value from /dev/urandom, as its a better source of entropy */
	fp = fopen("/dev/urandom", "r");
	if (fp != NULL) {
		seed = fgetc(fp);
		fclose(fp);
	}

	/* else fallback to using the current time as the seed */
	else
		seed = (int)time(NULL);

	srand(seed);
	for (x = 0; x < buffer_size; x++)
		buffer[x] = (int)'0' + (int)(72.0 * rand() / (RAND_MAX + 1.0));

	return;
}


/*
 * Convenience functions for exiting with a warning and error respectively.
 * Nagios ignores stderr and interprets return-code 2 as critical and 1 as
 * warning.
 */
void exit_crit(const char *fmt, ...)
{
	va_list ap;
	fprintf(stdout,"CRITICAL: ");
	va_start(ap, fmt);
	vfprintf(stdout, fmt, ap);
	va_end(ap);
	fprintf(stdout, "\n");
	exit(2);
}

void exit_warn(const char *fmt, ...)
{
	va_list ap;
	fprintf(stdout,"WARNING: ");
	va_start(ap, fmt);
	vfprintf(stdout, fmt, ap);
	va_end(ap);
	fprintf(stdout, "\n");
	exit(1);
}

void exit_unknown(const char *fmt, ...)
{
	va_list ap;
	fprintf(stdout,"UNKNOWN: ");
	va_start(ap, fmt);
	vfprintf(stdout, fmt, ap);
	va_end(ap);
	fprintf(stdout, "\n");
	exit(3);
}

/*
 * Open a connection to the hostname and port and store the FD on *sd.
 * Returns 0 on success. Iterates sockets until one works, thus
 * IPv4/IPv6-agnostic.
 */
int my_tcp_connect(char *host_name, int port, int *sd)
{
	struct addrinfo hints;
	struct addrinfo *result, *rp;
	int sfd, s;
	char strport[10];

	s = snprintf(strport,10,"%d",port);
	if (s>10 || s<0)
		exit_crit("Invalid port-number?");
	
	memset(&hints, 0, sizeof(struct addrinfo));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_flags = 0;
	hints.ai_protocol = 0;

	s = getaddrinfo(host_name, strport, &hints, &result);
	if (s != 0)
		exit_crit("getaddrinfo: %s\n", gai_strerror(s));

	for (rp = result; rp != NULL; rp = rp->ai_next) {
		sfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
		if (sfd == -1)
			continue;

		if (connect(sfd, rp->ai_addr, rp->ai_addrlen) != -1)
			break;

		close(sfd);
	}

	if (rp == NULL)
		exit_crit("Could not connect\n");

	freeaddrinfo(result);

	*sd = sfd;

	return 0;
}

/* sends all data - thanks to Beej's Guide to Network Programming */
int sendall(int s, char *buf, int *len)
{
	int total = 0;
	int bytesleft = *len;
	int n = 0;

	/* send all the data */
	while (total < *len) {

		/* send some data */
		n = send(s, buf + total, bytesleft, 0);

		/* break on error */
		if (n == -1)
			break;

		/* apply bytes we sent */
		total += n;
		bytesleft -= n;
	}

	/* return number of bytes actually send here */
	*len = total;

	/* return -1 on failure, 0 on success */
	return n == -1 ? -1 : 0;
}

/*
 * receives all data - modelled after sendall()
 *
 * FIXME: Should use SO_RCVTIMEOUT if applicable.
 * The sleep(1)-way is inherently slow and wasteful. If we're going to
 * wait, we might as well block.
 */
int recvall(int s, char *buf, int *len, int timeout)
{
	int total = 0;
	int bytesleft = *len;
	int n = 0;
	time_t start_time;
	time_t current_time;

	/* clear the receive buffer */
	bzero(buf, *len);

	time(&start_time);

	/* receive all data */
	while (total < *len) {

		/* receive some data */
		n = recv(s, buf + total, bytesleft, 0);

		/* no data has arrived yet (non-blocking socket) */
		if (n == -1 && errno == EAGAIN) {
			time(&current_time);
			if (current_time - start_time > timeout)
				break;
			sleep(1);
			continue;
		}

		/* receive error or client disconnect */
		else if (n <= 0)
			break;

		/* apply bytes we received */
		total += n;
		bytesleft -= n;
	}

	/* return number of bytes actually received here */
	*len = total;

	/* return <=0 on failure, bytes received on success */
	return (n <= 0) ? n : total;
}

/* fixes compiler problems under Solaris, since strsep() isn't included */
/* this code is taken from the glibc source */
char *my_strsep(char **stringp, const char *delim)
{
	char *begin, *end;

	begin = *stringp;
	if (begin == NULL)
		return NULL;

	/* A frequent case is when the delimiter string contains only one
	   character.  Here we don't need to call the expensive `strpbrk'
	   function and instead work using `strchr'.  */
	if (delim[0] == '\0' || delim[1] == '\0') {
		char ch = delim[0];

		if (ch == '\0')
			end = NULL;
		else {
			if (*begin == ch)
				end = begin;
			else
				end = strchr(begin + 1, ch);
		}
	}

	else
		/* Find the end of the token.  */
		end = strpbrk(begin, delim);

	if (end) {

		/* Terminate the token and set *STRINGP past NUL character.  */
		*end++ = '\0';
		*stringp = end;
	} else
		/* No more delimiters; this is the last token.  */
		*stringp = NULL;

	return begin;
}

/* show license */
void display_license(void)
{

	printf
	    ("This program is released under the GPL (see below) with the additional\n");
	printf
	    ("exemption that compiling, linking, and/or using OpenSSL is allowed.\n\n");

	printf
	    ("This program is free software; you can redistribute it and/or modify\n");
	printf
	    ("it under the terms of the GNU General Public License as published by\n");
	printf
	    ("the Free Software Foundation; either version 2 of the License, or\n");
	printf("(at your option) any later version.\n\n");
	printf
	    ("This program is distributed in the hope that it will be useful,\n");
	printf
	    ("but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
	printf
	    ("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
	printf("GNU General Public License for more details.\n\n");
	printf
	    ("You should have received a copy of the GNU General Public License\n");
	printf("along with this program; if not, write to the Free Software\n");
	printf("Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\n");

	return;
}


src/check_nrpe.c
/*
 * check_nrpe.c - NRPE Plugin For Nagios
 * Copyright (c) 1999-2008 Ethan Galstad (nagios@nagios.org)
 * Copyright (c) 2011 Kristian Lyngstol <kristian@bohemians.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

/*
 * XXX: I've inserted the GPLv2+ header as it was missing and only "GPL"
 * XXX: was mentioned. This was chosen as the GPLv2+-header is present
 * XXX: elsewhere in nrpe. It is presumed to apply to the existing code,
 * XXX: and it applies to my contributions as well.
 * XXX:    - Kristian Lyngstol, March, 2011
 *
 * This plugin will attempt to connect to the NRPE daemon on the specified server and port.
 * The daemon will attempt to run the command defined as [command].  Program output and
 * return code are sent back from the daemon and displayed as this plugin's own output and
 * return code.
 *
 * Command line: CHECK_NRPE -H <host_address> [-p port] [-c command] [-to to_sec]
 *
 */

#include <arpa/inet.h>
#include <netdb.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>

#include "common.h"
#include "config.h"
#include "utils.h"

#define DEFAULT_NRPE_COMMAND	"_NRPE_CHECK"	/* check version of NRPE daemon */

#ifdef HAVE_SSL
#include <openssl/ssl.h>

struct ssl_data {
	SSL_METHOD *meth;
	SSL_CTX *ctx;
	SSL *ssl;
};
#endif

struct check_nrpe {
	int server_port;
	char *server_name;
	char *command_name;
	int socket_timeout;
	int timeout_return_code;
	int sd;

	char query[MAX_INPUT_BUFFER];

	int show_help;
	int show_license;
	int show_version;

	int use_ssl;
#ifdef HAVE_SSL
	struct ssl_data ssl;
#endif
};

struct check_nrpe global;

/*
 * Just sets the global data structure to something sane.
 */
static void initialize_global(void) {
	global.server_port = DEFAULT_SERVER_PORT;
	global.server_name = NULL;
	global.command_name = NULL;
	global.socket_timeout = DEFAULT_SOCKET_TIMEOUT;
	global.timeout_return_code = STATE_CRITICAL;
	memset(global.query,sizeof(global.query),0);
	global.show_help = FALSE;
	global.show_license = FALSE;
	global.show_version = FALSE;
#ifdef HAVE_SSL
	global.use_ssl = TRUE;
#else
	global.use_ssl = FALSE;
#endif
}

static void usage(void)
{
	printf ("Usage: check_nrpe -H <host> [-n] [-u] [-p <port>] [-t <timeout>] [-c <command>] [-a <arglist...>]\n");
	printf("\n");
	printf("Options:\n");
	printf(" -n         = Do no use SSL\n");
	printf (" -u         = Make socket timeouts return an UNKNOWN state instead of CRITICAL\n");
	printf (" <host>     = The address of the host running the NRPE daemon\n");
	printf (" [port]     = The port on which the daemon is running (default=%d)\n",
		 DEFAULT_SERVER_PORT);
	printf (" [timeout]  = Number of seconds before connection times out (default=%d)\n",
		 DEFAULT_SOCKET_TIMEOUT);
	printf (" [command]  = The name of the command that the remote daemon should run\n");
	printf (" [arglist]  = Optional arguments that should be passed to the command.  Multiple\n");
	printf ("              arguments should be separated by a space.  If provided, this must be\n");
	printf ("              the last option supplied on the command line.\n");
	printf("\n");
	printf("Note:\n");
	printf ("This plugin requires that you have the NRPE daemon running on the remote host.\n");
	printf ("You must also have configured the daemon to associate a specific plugin command\n");
	printf ("with the [command] option you are specifying here.  Upon receipt of the\n");
	printf ("[command] argument, the NRPE daemon will run the appropriate plugin command and\n");
	printf ("send the plugin output and return code back to *this* plugin.  This allows you\n");
	printf ("to execute plugins on remote hosts and 'fake' the results to make Nagios think\n");
	printf("the plugin is being run locally.\n");
	printf("\n");
}

#define OPTCHARS "H:c:a:t:p:nuhl"
static int process_arguments(int argc, char **argv)
{
	int argindex = 0;
	int c = 1;
	int i = 1;

#ifdef HAVE_GETOPT_LONG
	int option_index = 0;
	static struct option long_options[] = {
		{"host", required_argument, 0, 'H'},
		{"command", required_argument, 0, 'c'},
		{"args", required_argument, 0, 'a'},
		{"no-ssl", no_argument, 0, 'n'},
		{"unknown-timeout", no_argument, 0, 'u'},
		{"timeout", required_argument, 0, 't'},
		{"port", required_argument, 0, 'p'},
		{"help", no_argument, 0, 'h'},
		{"license", no_argument, 0, 'l'},
		{0, 0, 0, 0}
	};
#endif

	/* no options were supplied */
	if (argc < 2)
		return ERROR;

	while (1) {
#ifdef HAVE_GETOPT_LONG
		c = getopt_long(argc, argv, OPTCHARS, long_options,
				&option_index);
#else
		c = getopt(argc, argv, OPTCHARS);
#endif
		if (c == -1 || c == EOF)
			break;

		/* process all arguments */
		switch (c) {

		case '?':
		case 'h':
			global.show_help = TRUE;
			break;
		case 'V':
			global.show_version = TRUE;
			break;
		case 'l':
			global.show_license = TRUE;
			break;
		case 't':
			global.socket_timeout = atoi(optarg);
			if (global.socket_timeout <= 0)
				return ERROR;
			break;
		case 'p':
			global.server_port = atoi(optarg);
			if (global.server_port <= 0)
				return ERROR;
			break;
		case 'H':
			global.server_name = strdup(optarg);
			break;
		case 'c':
			global.command_name = strdup(optarg);
			break;
		case 'a':
			argindex = optind;
			break;
		case 'n':
			global.use_ssl = FALSE;
			break;
		case 'u':
			global.timeout_return_code = STATE_UNKNOWN;
			break;
		default:
			return ERROR;
			break;
		}
	}

	/* FIXME: Check size properly instead of cutting off the end */
	snprintf(global.query, sizeof(global.query), "%s",
		 (global.command_name == NULL) ? DEFAULT_NRPE_COMMAND : global.command_name);
	global.query[sizeof(global.query) - 1] = '\x0';

	/* get the command args */
	if (argindex > 0) {

		for (c = argindex - 1; c < argc; c++) {

			i = sizeof(global.query) - strlen(global.query) - 2;
			if (i <= 0)
				break;

			strcat(global.query, "!");
			strncat(global.query, argv[c], i);
			global.query[sizeof(global.query) - 1] = '\x0';
		}
	}

	/* make sure required args were supplied */
	if (global.server_name == NULL && global.show_help == FALSE && global.show_version == FALSE
	    && global.show_license == FALSE)
		return ERROR;

	return OK;
}

static void alarm_handler(int __attribute__((unused)) sig)
{
	printf("CHECK_NRPE: Socket timeout after %d seconds.\n",
	       global.socket_timeout);

	exit(global.timeout_return_code);
}

/* submitted by Mark Plaksin 08/31/2006 */
static int graceful_close(int sd, int timeout)
{
	fd_set in;
	struct timeval tv;
	char buf[1000];

	/* send FIN packet */
	shutdown(sd, SHUT_WR);
	for (;;) {

		FD_ZERO(&in);
		FD_SET(sd, &in);
		tv.tv_sec = timeout / 1000;
		tv.tv_usec = (timeout % 1000) * 1000;

		/* timeout or error */
		if (1 != select(sd + 1, &in, NULL, NULL, &tv))
			break;

		/* no more data (FIN or RST) */
		if (0 >= recv(sd, buf, sizeof(buf), 0))
			break;
	}

#ifdef HAVE_CLOSESOCKET
	closesocket(sd);
#else
	close(sd);
#endif

	return OK;
}
int main(int argc, char **argv)
{
	uint32_t packet_crc32;
	uint32_t calculated_crc32;
	int16_t result;
	int rc;
	packet send_packet;
	packet receive_packet;
	int bytes_to_send;
	int bytes_to_recv;

	initialize_global();
	result = process_arguments(argc, argv);

	if (result != OK || global.show_help == TRUE || global.show_license == TRUE
	    || global.show_version == TRUE) {

		if (result != OK)
			printf("Incorrect command line arguments supplied\n");
		printf("\n");
		printf("NRPE Plugin for Nagios\n");
		printf
		    ("Copyright (c) 1999-2008 Ethan Galstad (nagios@nagios.org)\n");
		printf("Version: " PACKAGE_VERSION "\n");
		printf("License: GPL v2 with exemptions (-l for more info)\n");
#ifdef HAVE_SSL
		printf
		    ("SSL/TLS Available: Anonymous DH Mode, OpenSSL 0.9.6 or higher required\n");
#endif
		printf("\n");
	}

	if (result != OK || global.show_help == TRUE)
		usage();	

	if (global.show_license == TRUE)
		display_license();

	if (result != OK || global.show_help == TRUE || global.show_license == TRUE
	    || global.show_version == TRUE)
		exit(STATE_UNKNOWN);

	/* generate the CRC 32 table */
	generate_crc32_table();

#ifdef HAVE_SSL
	/* initialize SSL */
	if (global.use_ssl == TRUE) {
		SSL_library_init();
		SSLeay_add_ssl_algorithms();
		global.ssl.meth = SSLv23_client_method();
		SSL_load_error_strings();
		if ((global.ssl.ctx = SSL_CTX_new(global.ssl.meth)) == NULL)
			exit_crit("Could not create SSL context.");

		/* ADDED 01/19/2004 */
		/* use only TLSv1 protocol */
		SSL_CTX_set_options(global.ssl.ctx, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
	}
#endif

	/* initialize alarm signal handling */
	signal(SIGALRM, alarm_handler);

	/* set socket timeout */
	alarm(global.socket_timeout);

	/* try to connect to the host at the given port number */
	result = my_tcp_connect(global.server_name, global.server_port, &global.sd);

#ifdef HAVE_SSL
	/* do SSL handshake */
	if (result == STATE_OK && global.use_ssl == TRUE) {
		if ((global.ssl.ssl = SSL_new(global.ssl.ctx)) != NULL) {
			SSL_CTX_set_cipher_list(global.ssl.ctx, "ADH");
			SSL_set_fd(global.ssl.ssl, global.sd);
			if ((rc = SSL_connect(global.ssl.ssl)) != 1)
				exit_crit("Could not complete SSL handshake.");
		} else {
			exit_crit("Could not create SSL connection structure.");
		}

		/* bail if we had errors */
		if (result != STATE_OK) {
			SSL_CTX_free(global.ssl.ctx);
			close(global.sd);
			exit(result);
		}
	}
#endif

	/* we're connected and ready to go */
	if (result == STATE_OK) {

		/* clear the packet buffer */
		bzero(&send_packet, sizeof(send_packet));

		/* fill the packet with semi-random data */
		randomize_buffer((char *)&send_packet, sizeof(send_packet));

		/* initialize packet data */
		send_packet.packet_version =
		    (int16_t) htons(NRPE_PACKET_VERSION_2);
		send_packet.packet_type = (int16_t) htons(QUERY_PACKET);
		strncpy(&send_packet.buffer[0], global.query, MAX_PACKETBUFFER_LENGTH);
		send_packet.buffer[MAX_PACKETBUFFER_LENGTH - 1] = '\x0';

		/* calculate the crc 32 value of the packet */
		send_packet.crc32_value = (u_int32_t) 0L;
		calculated_crc32 =
		    calculate_crc32((char *)&send_packet, sizeof(send_packet));
		send_packet.crc32_value = (u_int32_t) htonl(calculated_crc32);

		/***** ENCRYPT REQUEST *****/

		/* send the packet */
		bytes_to_send = sizeof(send_packet);
		if (global.use_ssl == FALSE)
			rc = sendall(global.sd, (char *)&send_packet, &bytes_to_send);
#ifdef HAVE_SSL
		else {
			rc = SSL_write(global.ssl.ssl, &send_packet, bytes_to_send);
			if (rc < 0)
				rc = -1;
		}
#endif
		/*
		 * FIXME: Should be unknown, not crit.
		 */
		if (rc == -1) {
			close(global.sd);
			exit_crit("Error sending query to host.");
		}

		/* wait for the response packet */
		bytes_to_recv = sizeof(receive_packet);
		if (global.use_ssl == FALSE)
			rc = recvall(global.sd, (char *)&receive_packet,
				     &bytes_to_recv, global.socket_timeout);
#ifdef HAVE_SSL
		else
			rc = SSL_read(global.ssl.ssl, &receive_packet, bytes_to_recv);
#endif

		/* reset timeout */
		alarm(0);

		/* close the connection */
#ifdef HAVE_SSL
		if (global.use_ssl == TRUE) {
			SSL_shutdown(global.ssl.ssl);
			SSL_free(global.ssl.ssl);
			SSL_CTX_free(global.ssl.ctx);
		}
#endif
		graceful_close(global.sd, 1000);

		if (rc < 0)
			exit_unknown("Error receiving data from daemon.");
		else if (rc == 0)
			exit_unknown("Received 0 bytes from daemon.  Check the remote server logs for error messages.");
		else if (bytes_to_recv < sizeof(receive_packet))
			exit_unknown("Receive underflow - only %d bytes received (%lu expected).",
			     bytes_to_recv, sizeof(receive_packet));

		/***** DECRYPT RESPONSE *****/

		/* check the crc 32 value */
		packet_crc32 = ntohl(receive_packet.crc32_value);
		receive_packet.crc32_value = 0L;
		calculated_crc32 =
		    calculate_crc32((char *)&receive_packet,
				    sizeof(receive_packet));
		if (packet_crc32 != calculated_crc32) {
			close(global.sd);
			exit_unknown("Response packet had invalid CRC32.");
		}

		/* check packet version */
		if (ntohs(receive_packet.packet_version) !=
			NRPE_PACKET_VERSION_2) {
			close(global.sd);
			exit_unknown("Invalid packet version received from server.");
		}

		/* check packet type */
		if (ntohs(receive_packet.packet_type) != RESPONSE_PACKET) {
			close(global.sd);
			exit_unknown("Invalid packet type received from server.");
		}

		/* get the return code from the remote plugin */
		result = (int16_t) ntohs(receive_packet.result_code);

		/* print the output returned by the daemon */
		receive_packet.buffer[MAX_PACKETBUFFER_LENGTH - 1] = '\x0';
		if (!strcmp(receive_packet.buffer, ""))
			printf("CHECK_NRPE: No output returned from daemon.\n");
		else
			printf("%s\n", receive_packet.buffer);
	}

	/* reset the alarm */
	else
		alarm(0);

	return result;
}

